File: src-tauri/src/agent/protocol.rs
use serde::{Deserialize, Serialize};

/// Agent Protocol 版本。Rust 结构体为真相源；Python 用 pydantic 镜像，TS 由 specta/tauri-typegen 生成。
pub const PROTOCOL_VERSION: u8 = 1;

/// 所有协议消息的统一信封（每个帧必备基础字段）。
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Envelope {
    pub protocol_version: u8,
    pub id: String,
    pub run_id: String,
    pub session_id: String,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub created_at: String,
    pub payload: serde_json::Value,
}

impl Envelope {
    /// 构造一个回包信封（V0.1 骨架：握手用，created_at 留空，会话/运行上下文在 V0.2 填充）。
    pub fn reply(msg_type: &str, payload: serde_json::Value) -> Self {
        Envelope {
            protocol_version: PROTOCOL_VERSION,
            id: uuid::Uuid::new_v4().to_string(),
            run_id: String::new(),
            session_id: String::new(),
            msg_type: msg_type.to_string(),
            created_at: String::new(),
            payload,
        }
    }
}

/// 消息类型（方向在架构文档 §3.2 已定死）：
/// Rust→Python: run_request / tool_result / approval_result / run_cancel / user_message
/// Python→Rust: assistant_delta / tool_call_request / memory_query_request / run_finished / run_error
/// 双向握手: hello / ready / ping / pong
pub mod msg_type {
    pub const HELLO: &str = "hello";
    pub const READY: &str = "ready";
    pub const RUN_REQUEST: &str = "run_request";
    pub const TOOL_CALL_REQUEST: &str = "tool_call_request";
    pub const TOOL_RESULT: &str = "tool_result";
    pub const APPROVAL_RESULT: &str = "approval_result";
    pub const RUN_CANCEL: &str = "run_cancel";
    pub const RUN_FINISHED: &str = "run_finished";
    pub const RUN_ERROR: &str = "run_error";
    pub const ASSISTANT_DELTA: &str = "assistant_delta";
    pub const MEMORY_QUERY_REQUEST: &str = "memory_query_request";
    pub const DEBUG_PROMPT: &str = "debug_prompt";
    pub const DEBUG_PROMPT_RESULT: &str = "debug_prompt_result";
    pub const EMBEDDING_REQUEST: &str = "embedding_request";
    pub const EMBEDDING_RESULT: &str = "embedding_result";
    pub const PING: &str = "ping";
    pub const PONG: &str = "pong";
}

File: agent/app/protocol.py
"""Agent Protocol 消息（pydantic 镜像 Rust 真相源，见 src-tauri/src/agent/protocol.rs）。"""
from __future__ import annotations

import time
import uuid
from enum import Enum

from pydantic import BaseModel

PROTOCOL_VERSION = 1


class MsgType(str, Enum):
    HELLO = "hello"
    READY = "ready"
    PING = "ping"
    PONG = "pong"
    RUN_REQUEST = "run_request"
    TOOL_CALL_REQUEST = "tool_call_request"
    TOOL_RESULT = "tool_result"
    APPROVAL_RESULT = "approval_result"
    RUN_CANCEL = "run_cancel"
    RUN_FINISHED = "run_finished"
    RUN_ERROR = "run_error"
    ASSISTANT_DELTA = "assistant_delta"
    MODEL_FALLBACK = "model_fallback"
    MEMORY_QUERY_REQUEST = "memory_query_request"
    USER_MESSAGE = "user_message"
    DEBUG_PROMPT = "debug_prompt"
    DEBUG_PROMPT_RESULT = "debug_prompt_result"
    EMBEDDING_REQUEST = "embedding_request"
    EMBEDDING_RESULT = "embedding_result"


class Envelope(BaseModel):
    protocol_version: int = PROTOCOL_VERSION
    id: str = ""
    run_id: str = ""
    session_id: str = ""
    type: str
    created_at: str = ""
    payload: dict = {}


def make(
    msg_type: str,
    session_id: str = "",
    run_id: str = "",
    payload: dict | None = None,
) -> Envelope:
    return Envelope(
        id=uuid.uuid4().hex,
        run_id=run_id,
        session_id=session_id,
        type=msg_type,
        created_at=time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        payload=payload or {},
    )
